### PR TITLE
[JENKINS-46834] Add flexible inheritance strategies, deprecate 'blocksInheritance'

### PR DESCRIPTION
**Pull request by daniel-beck**
I believe this addresses the following issues in [JENKINS-46834](https://issues.jenkins-ci.org/browse/JENKINS-46834):

* [JENKINS-39873](https://issues.jenkins-ci.org/browse/JENKINS-39873), 
* [JENKINS-37904](https://issues.jenkins-ci.org/browse/JENKINS-37904), 
* [JENKINS-39409](https://issues.jenkins-ci.org/browse/JENKINS-39409), 
* [JENKINS-38356](https://issues.jenkins-ci.org/browse/JENKINS-38356), 
* possibly [JENKINS-36788](https://issues.jenkins-ci.org/browse/JENKINS-36788), 
* [JENKINS-24878](https://issues.jenkins-ci.org/browse/JENKINS-24878) (sort of),

Incidentally as part of cleanup, [JENKINS-33315](https://issues.jenkins-ci.org/browse/JENKINS-33315) is also believed to be addressed.

Supersedes https://github.com/jenkinsci/matrix-auth-plugin/pull/15.

I may rewrite history.

screen shot

@jenkinsci/code-reviewers

### CODE REVIEW

File: src/main/java/hudson/security/GlobalMatrixAuthorizationStrategy.java [hunk 1]
```diff
@@ -412,8 +412,18 @@ protected GlobalMatrixAuthorizationStrategy create() {
         }
 
         public List<PermissionGroup> getAllGroups() {
-            List<PermissionGroup> groups = new ArrayList<PermissionGroup>(PermissionGroup.getAll());
-            groups.remove(PermissionGroup.get(Permission.class));
+            List<PermissionGroup> groups = new ArrayList<PermissionGroup>();
```
Comment (daniel-beck) [inline src/main/java/hudson/security/GlobalMatrixAuthorizationStrategy.java]: This is the fix for JENKINS-33315.

File: src/main/java/org/jenkinsci/plugins/matrixauth/AuthorizationProperty.java [hunk 2]
```diff
@@ -99,6 +108,9 @@ default boolean hasPermission(String sid, Permission p, boolean principal) {
      * Checks if the permission is explicitly given, instead of implied through {@link Permission#impliedBy}.
      */
     default boolean hasExplicitPermission(String sid, Permission p) {
+        if (sid == null) { // used for template row in UI
```
Comment (daniel-beck) [inline src/main/java/org/jenkinsci/plugins/matrixauth/AuthorizationProperty.java]: This fixes another occurrence of JENKINS-46190.

File: src/main/java/org/jenkinsci/plugins/matrixauth/AuthorizationProperty.java [hunk 3]
```diff
@@ -144,4 +156,8 @@ default void add(String shortForm) {
         return Arrays.asList(data);
     }
 
+    @Restricted(NoExternalUse.class)
+    default ExtensionList<InheritanceStrategyDescriptor> getInheritanceStrategies() {
```
Comment (daniel-beck) [inline src/main/java/org/jenkinsci/plugins/matrixauth/AuthorizationProperty.java]: Unused.

File: src/main/resources/org/jenkinsci/plugins/matrixauth/AuthorizationMatrixNodeProperty/config.groovy [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+import lib.FormTagLib
+import org.jenkinsci.plugins.matrixauth.inheritance.InheritanceStrategyDescriptor
+
+def f = namespace(FormTagLib)
+def st = namespace("jelly:stapler")
+
+f.nested {
+    table(style: "width: 100%") {
+        f.dropdownDescriptorSelector(title: _("Inheritance Strategy"), descriptors: InheritanceStrategyDescriptor.getApplicableDescriptors(my.class), field: 'inheritanceStrategy')
+//        f.optionalBlock(field: "blocksInheritance", title:_("Block inheritance of global authorization matrix"))
```
Comment (daniel-beck) [inline src/main/resources/org/jenkinsci/plugins/matrixauth/AuthorizationMatrixNodeProperty/config.groovy]: Still need to remove this, obsolete.

**Comment by daniel-beck** [comment]
I probably need to restore full support for the `blocksInheritance` setters/getters and map those to/from best matching InheritanceStrategies. Too likely to break things otherwise.

File: src/main/java/hudson/security/ProjectMatrixAuthorizationStrategy.java [hunk 5]
```diff
@@ -88,20 +87,23 @@ public ACL getACL(ItemGroup g) {
         return getRootACL();
     }
 
+    @Nonnull
+    @Override
+    public ACL getACL(@Nonnull Node node) {
```
Comment (daniel-beck) [inline src/main/java/hudson/security/ProjectMatrixAuthorizationStrategy.java]: Like https://github.com/jenkinsci/matrix-auth-plugin/pull/27/commits/e69047a9c6cabfb6d5805d86a5509b4308392891 but adapted for inheritance strategies.

File: pom.xml [hunk 6]
```diff
@@ -31,14 +31,9 @@
         <dependency>
             <groupId>org.jenkins-ci.plugins</groupId>
             <artifactId>cloudbees-folder</artifactId>
-            <version>6.0.4</version>
+            <version>6.1.0</version>
             <optional>true</optional>
         </dependency>
-        <dependency> <!-- TODO JENKINS-33095 workaround -->
-            <groupId>org.jenkins-ci.plugins.icon-shim</groupId>
-            <artifactId>icon-shim</artifactId>
```
Comment (jglick) [inline pom.xml]: I am afraid I do not understand the policy around `icon-shim` at all. Cf. https://github.com/jenkinsci/plugin-pom/pull/72.

**Pull request merged**